zerver/webhooks/jira/view.py

```diff
@@ -81,20 +85,63 @@ def convert_jira_markup(content: str, realm: Realm) -> str:
     # Try to convert a Jira user mention of format [~username] into a
     # Zulip user mention. We don't know the email, just the Jira username,
     # so we naively guess at their Zulip account using this
-    mention_re = re.compile(r"\[~(.*?)\]")
-    for username in mention_re.findall(content):
-        # Try to look up username
-        user_profile = guess_zulip_user_from_jira(username, realm)
-        if user_profile:
-            replacement = f"**{user_profile.full_name}**"
-        else:
-            replacement = f"**{username}**"
-
-        content = content.replace(f"[~{username}]", replacement)
+    account_ids = get_mentioned_account_ids(content)
+    if len(account_ids) > 0 and auth is not None:
+        url = get_url(payload)
+        usernames = get_usernames_from_account_ids(url, account_ids, auth)
+        for username in usernames:
+            # Try to look up username
+            user_profile = guess_zulip_user_from_jira(username, realm)
+            if user_profile:
+                replacement = f"**{user_profile.full_name}**"
+            else:
+                replacement = f"**{username}**"
+
+            content = content.replace(f"[~{username}]", replacement)
 
     return content
 
 
+def get_mentioned_account_ids(content: str) -> list[str]:
+    mention_re = re.compile(r"\[\~accountid:([\w\-:]+)\]")
+    return mention_re.findall(content)
+
+
+def replace_account_ids_with_usernames(comment: str, usernames: list[str]) -> str:
+    pattern = r"(\[~accountid:[\w\-:]+\])"
+    parts = re.split(pattern, comment)
+    splitted_list = [part for part in parts if part]
+    username_index = 0  # Index of username to be replaced
+    for i, word in enumerate(splitted_list):
+        if re.match(pattern, word) is not None:
+            splitted_list[i] = f"**{usernames[username_index]}**"
+            username_index += 1
+    return "".join(splitted_list)
+
+
+def get_usernames_from_account_ids(
+    url: str, account_ids: list[str], auth: tuple[str, str]
+) -> list[str]:
+    usernames = []
+    for account_id in account_ids:
+        username = get_mentioned_account_name(url, auth, account_id)
+        usernames.append(username)
+    return usernames
+
+
+def get_mentioned_account_name(url: str, auth: tuple[str, str], account_id: str) -> str:
+    r = requests.get(f"{url}?accountId={account_id}", auth=auth)
+    res = json.loads(r.text)["displayName"]
```

Comment:
I think we want to do additional checks before assuming we got the response we want and handling those cases properly.

For reference:
```Python
def get_slack_api_data(slack_api_url: str, get_param: str, **kwargs: Any) -> Any:
    if not kwargs.get("token"):
        raise AssertionError("Slack token missing in kwargs")
    token = kwargs.pop("token")
    data = requests.get(slack_api_url, headers={"Authorization": f"Bearer {token}"}, params=kwargs)

    if data.status_code == requests.codes.ok:
        result = data.json()
        if not result["ok"]:
            raise Exception("Error accessing Slack API: {}".format(result["error"]))
        return result[get_param]

    raise Exception("HTTP error accessing the Slack API.")
```
One thing to note is, here, we need to take care to not throw Exception (error 500), instead we should probably return error 401 or 400 in the case where something unexpected happens.